NF: Automatic bvecs correction using fiber coherence index

CHrlS98 opened the pull request:
Hi,
New script for detecting bvecs flips/permutations and generating a corrected bvecs file: `scil_correct_gradients.py`. The script detects the combination of axis permutations/flips maximizing a fiber coherence index as described in *Schilling et al, 2019* and applies the transform maximizing fiber coherence to the gradients table. If the input bvecs are already correct, it will simply save the input.

See the documentation of the script for an example on how to use it. Takes less than 1 minute to run on a BIL&GIN subject.

I also updated the scripts `scil_flip_gradients.py` and `scil_swap_gradients_axis.py` to support additional bvecs file types. Does it look alright?

To do: 
- [x] Add tests for `scil_correct_gradients.py`;
- [x] Update tests for `scil_swap_gradients_axis.py` and `scil_flip_gradients.py`;
- [x] Fix output shape for `scil_swap_gradients_axis.py` and `scil_flip_gradients.py`;

**Should we add gradient correction as an optional step in `scil_compute_dti_metrics.py`?**

mdesco commented:
@CHrlS98 can you provide a little data test set we can download to test and play with the PR please?

CHrlS98 commented:
@mdesco a dataset with the `bvec` the original bvecs and `bvec_flipx_swapyz` the bvecs with a sign change along the X axis and a swap of the Y and Z axes. https://usherbrooke-my.sharepoint.com/:u:/g/personal/poic2312_usherbrooke_ca/EXfWZTy4CvpLv6LKL4rtrZkBpMcCG2n41uM5U2QI0zjAxA?e=YWS26n 

The link should work as long as you use your usherbrooke mail address. You can use the flip and swap scripts to generate other corrupted gradients table. I like using the file `mask.nii.gz`, which is simply a thresholded b0, to exclude the background when computing DTI metrics. Not necessary to use it again in `scil_correct_gradients.py` since black voxels are ignored anyway.

Possible issue, since I added support for other bvec file format in `scil_flip_gradients.py` and `scil_swap_gradients_axis.py`, the output bvec might be transposed with respect to the input bvec. I'll fix that.

CHrlS98 commented:
Here, I'll remove the WIP tag as the items I noted are now completed.

One question remains: **Should we add gradient correction as an optional step in `scil_compute_dti_metrics.py`?** I think the script as a standalone is cool and gives a change to analyse the corrected gradients before running the whole DTI metrics pipeline.

Let me know what you think!

mdesco commented on scripts/scil_correct_gradients.py:
```diff
@@ -0,0 +1,95 @@
+# -*- coding: utf-8 -*-
+"""
+Validate and correct b-vectors using a fiber coherence index as presented
+in Schilling et al, 2019. The script takes as input the principal direction
```
lets call the script scil_validate_and_correct_bvecs.py no?

CHrlS98 replied:
yes sounds good to me!

mdesco commented on scripts/scil_correct_gradients.py:
```diff
@@ -0,0 +1,95 @@
+A typical pipeline could be:
+>>> scil_compute_dti_metrics.py dwi.nii.gz bval bvec --not_all --fa fa.nii.gz
+    --evecs peaks.nii.gz
+>>> scil_correct_gradients.py bvec peaks_v1.nii.gz fa.nii.gz bvec_corr
```
Would it work if peaks come from a fODF? I can test. Should we mention something about this?

mdesco replied:
I have my answer, I just tested and it breaks...
 File "/Users/desm2239/Research/Source/scilpy/scilpy/reconst/fiber_coherence.py", line 43, in compute_fiber_coherence_table
    index = compute_fiber_coherence(directions.dot(t), values)
ValueError: shapes (136,184,132,15) and (3,3) not aligned: 15 (dim 3) != 3 (dim 0)

mdesco replied:
mrconvert peaks.nii.gz peaks_v1.nii.gz -coord 3 0:2 can do the trick... and just get the first peak of the fODF field.

CHrlS98 replied:
Yeah there needs to be only one peak par voxel, I will raise a parser error if the shape of the last axis of the 4D volume is different than 3.

mdesco replied:
Or should we run if first 3 floats? And just raise a warning?
That way it would work with fODFs peaks.

CHrlS98 replied:
I'll add a script to extract the peak associated to the maximum value for fODF, this way we can use the output as input for this script

CHrlS98 replied:
I did not add an additional script after all. I just added a step to extract the principal direction from a volume with more than one direction for each voxel, see fa38a53.

mdesco commented on scripts/scil_correct_gradients.py:
```diff
@@ -0,0 +1,95 @@
+>>> scil_correct_gradients.py bvec peaks_v1.nii.gz fa.nii.gz bvec_corr
+
+Note that peaks_v1.nii.gz is the file containing the direction associated
+to the highest eigen value at each voxel.
```
eigenvalue (one word)

mdesco commented on scripts/scil_correct_gradients.py:
```diff
@@ -0,0 +1,95 @@
+    p.add_argument('--mask', help='Path to an optional mask.')
+    p.add_argument('--fa_th', default=0.2, type=float,
+                   help='FA threshold. Only voxels with FA higher '
+                        'than fa_th will be considered.')
```
Please add the default value 0.2 in help. See other scripts.

mdesco reviewed (commented):
Works really well. Just address the few edits and I will approve.

mdesco reviewed (approved):
Nice job! Ready to be merged.

CHrlS98 commented:
It would appear multiple peaks suport does not work when directly using eigenvectors and eigenvalues taken directly from `scil_compute_dti_metrics.py`. I'll fix this soon.

arnaudbore commented on scripts/scil_validate_and_correct_bvecs.py:
```diff
@@ -0,0 +1,119 @@
+    p = argparse.ArgumentParser(formatter_class=argparse.RawTextHelpFormatter,
+                                description=__doc__)
+
+    p.add_argument('in_bvecs', help='Path to bvecs file.')
```
please put the help on a second line. Just for homogeneity. Check also others script. 
Example: in_bvec instead of in_bvecs, in_FA instead of in_fa.

arnaudbore reviewed (commented):
Quick message, following our discussion IRL. Split your modifications related to gradient "dipy" format. Please add --dipy option to flip, swap, visualise scripts and explain why you need this extension. Few minor modifications.

CHrlS98 commented:
@arnaudbore 

> Quick message, following our discussion IRL. Split your modifications related to gradient "dipy" format. Please add --dipy option to flip, swap, visualise scripts and explain why you need this extension. Few minor modifications.

The issue I was describing wasn't a real issue. My bvecs were in FSL format, they were simply missing the `.bvec` required extension. Hence `dipy` gradient format isn't of any use.

I reverted all changes to flip/swap scripts and won't do a separate review for them.
